src/v/pandaproxy/schema_registry/types.h

```diff
@@ -59,21 +56,39 @@ from_string_view<schema_type>(std::string_view sv) {
       .default_match(std::nullopt);
 }
 
-inline std::ostream& operator<<(std::ostream& os, const schema_type& v) {
-    return os << to_string_view(v);
-}
+std::ostream& operator<<(std::ostream& os, const schema_type& v);
 
 ///\brief A subject is the name under which a schema is registered.
 ///
 /// Typically it will be "<topic>-key" or "<topic>-value".
 using subject = named_type<ss::sstring, struct subject_tag>;
 static const subject invalid_subject{};
 
+///\brief An unvalidated definition of the schema and its type.
+struct raw_schema_definition {
+    using str = named_type<ss::sstring, struct str_tag>;
+
+    template<typename T>
+    raw_schema_definition(T&& def, schema_type type)
+      : def{ss::sstring{std::forward<T>(def)}}
+      , type{type} {}
+
+    bool friend operator==(
+      const raw_schema_definition& lhs, const raw_schema_definition& rhs)
+      = default;
+
+    friend std::ostream&
+    operator<<(std::ostream& os, const raw_schema_definition& def);
+
+    str def;
```

Comment:
instead of doing `def.def()` you can use `const ss::string& operator() const { return def();}` and have less code churn.

Reply:
There are multiple states I'm trying to unravel:
* `raw` - an unparsed schema definition - usually from the user.
* `sanitized` - should be the canonical form; for Avro this is currently minified JSON (but should be [Parsing Canonical Form](https://avro.apache.org/docs/current/spec.html#Parsing+Canonical+Form+for+Schemas))
* `valid` - This has been checked along with its references (this is state isn't represented as a type yet)

Then there are various native types, which are built, and cross-referenced:
* `avro::ValidSchema`
* `google::protobuf::FileDescriptor*`
* Json Schema (not implemented yet)

So I'm, erring on the side of caution with conversions.

The PR for Protobuf support will make `schema_definition` a variant, something like this:

```c++
using schema_definition = std::variant<
  raw_schema_definition,
  avro_schema_definition,
  protobuf_schema_definition>;
```

Reply:
same with `sanitized` -> `canonical_`

Reply:
> if `raw` means `unparsed` why not name it `unparsed_schema_definition`

> same with `sanitized` -> `canonical_`

It's a good suggestion. I'll also introduce `canonical_schema_definition` in this PR.